Add Actor Categories and Categories dropdown filter in Map editor

rob-v opened the pull request:
Closes #10139

Add Actor categories:
V01:
..EditorTilesetFilter:
....ExcludeTilesets: DESERT
....**Categories: Building, Civilian**

Add Categories dropdown filter in Map editor:

![obrazok](https://cloud.githubusercontent.com/assets/16348750/26562808/3a1a6a80-44ca-11e7-8490-8ea4afd17517.png)

Categories list is combined from all existing categories of all actors.

EditorTilesetFilter can be renamed but due to currently broken Utility, hundreds of changed files (due to renaming or removed comments), I prefer to do it in another PR or at least lately as final commit after review of this PR.

Commit:
Add Actor Categories and Category dropdown filter in Map editor

pchote commented:
#13344 fixes the utility.

reaperrr commented:
This works rather well, the only thing I found a bit inconvenient is that if the list of categories is rather long, toggling off all unwanted categories one by one is rather tedious. Would it be possible to add a "Select all" toggle that can toggle all categories on/off at once?

rob-v commented:
Updated - added Select all, Select none options

MustaphaTR reviewed (approved):
Works as described. :+1:

pchote commented on mods/cnc/chrome/editor.yaml:
```diff
@@ -409,3 +415,24 @@ Container@EDITOR_WORLD_ROOT:
+ScrollPanel@ACTOR_CATEGORY_FILTER_PANEL:
+	Width: PARENT_RIGHT
+	Children:
+		Checkbox@SELECT_ALL_CATEGORIES:
```
IMO these would be more intuitive as buttons on the same row.

pchote reviewed (dismissed):
All of the actors for all mods show up in the one "Not category" category.  Can you please set up `EditorTilesetFilter` traits on the defaults in each mod?

reaperrr commented:
@pchote What just crossed my mind, how compatible is this approach to categories wrt translation?

pchote commented:
For that it would be best if we split the internal ID away from the display name shown in the UI.

rob-v commented:
Updated - checkboxes replaced by buttons - see new image in PR's description

obrakmann reviewed (approved):
lgtm, 👍

rob-v commented:
Is there someone having idea/time to set up EditorTilesetFilter traits - Categories on the defaults in each mod? I probably won't work on this next week(s), so I would close it to free PR queue.

pchote commented:
I should think that grouping based on the templates in defaults.yaml should be good enough, and easy to do.

pchote commented:
That should only be a few minutes work, so it would be a shame to give up on this feature over it.

reaperrr commented:
@rob-v @pchote I've set up actor categories locally, I can either push them to this branch or we merge it first and I file a separate PR.

rob-v commented:
@reaperrr I talked with @pchote on IRC so was also just finishing it. really not too difficult and results looks useful. IMO if you have it ready, it can be only better than my version, so do as you prefer.

reaperrr commented:
Ah sorry, should have checked on IRC first.
If you have your changes ready, I'd say push them first, Travis didn't run on your last update so that way we could also make sure there are no rebase conflicts.

If there is anything in my categories that I consider an improvement, I can still file them in a follow-up.

rob-v commented:
Thanks for your willigness. I should have rather confirmed here that I'll finish it. Ok, so I'll push it soon.

rob-v commented:
Updated - added Actor Categories for all mods.

reaperrr commented on mods/ra/rules/misc.yaml:
```diff
@@ -208,6 +210,8 @@ MINE:
 		Terrain: Ore
 	SeedsResource:
+	EditorTilesetFilter:
+		Categories: Mine
```
This will put Ore and Gem mines in the same category as explosive mines. I suggest Resource spawns for the ore/gem/rail mines here.

pchote replied:
The "rail" mines don't spawn resources FYI

reaperrr commented on mods/cnc/rules/defaults.yaml:
```diff
@@ -563,6 +573,8 @@
 		Type: Circle
 			Radius: 427
+	EditorTilesetFilter:
+		Categories: Huntable
```
I think 'Critter' is more descriptive than Huntable (for visc, too).

rob-v commented:
Updated - Ore and Gem mines now 'Resource spawn', Huntable changed to Critter.

pchote commented on mods/cnc/rules/defaults.yaml:
```diff
@@ -827,6 +857,8 @@
 	FrozenUnderFog:
 	ScriptTriggers:
+	EditorTilesetFilter:
+		Categories: Husk
```
I'd personally include this under trees (or maybe both?).

pchote commented on mods/cnc/rules/misc.yaml:
```diff
@@ -76,6 +76,8 @@ CAMERA:
 		Range: 10c0
 		Type: CenterPosition
+	EditorTilesetFilter:
+		Categories: Camera
```
How about a `System` category that has these, waypoints, and start points?

pchote replied:
Likewise for the other mods, possibly including `flameguy` in TS and the spice blooms in D2K.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Editor/ActorSelectorLogic.cs:
```diff
@@ -75,6 +79,61 @@ public ActorSelectorLogic(Widget widget, World world, WorldRenderer worldRendere
+				.Where(i => i != null && i.Categories != null &&
+					!(i.ExcludeTilesets != null && i.ExcludeTilesets.Contains(tileSetId)) && !(i.RequireTilesets != null && !i.RequireTilesets.Contains(tileSetId)))
+				.SelectMany(i => i.Categories).Distinct().OrderBy(i => i).ToList();
+			allCategories.Insert(0, NoCategory);
```
It would be good to skip adding this if there aren't any uncategorized actors.

pchote commented on mods/common/chrome/editor.yaml:
```diff
@@ -394,3 +401,29 @@ Container@EDITOR_WORLD_ROOT:
+					Y: 2 
+					Width: 90
+					Height: 25
+					Text: Select all
```
These both should be `Font: Bold`

pchote commented on mods/ts/rules/defaults.yaml:
```diff
@@ -809,16 +835,22 @@
 	Tooltip:
 		Name: Box
+	EditorTilesetFilter:
+		Categories: Box
```
Better to leave these as Decoration, IMO.

pchote commented on mods/ts/rules/misc.yaml:
```diff
@@ -94,6 +96,8 @@ CRATE:
 		Condition: crate-speed
 		Notification: 00-i080.aud
+	EditorTilesetFilter:
+		Categories: Crate
```
Crates have a limited lifetime so can't really be placed on the map.  This makes them a good candidate for the "System" category I suggested above.

pchote commented on mods/ts/rules/defaults.yaml:
```diff
@@ -809,16 +835,22 @@
 	Tooltip:
 		Name: Drum
+	EditorTilesetFilter:
+		Categories: Drum
```
These are best left as Decoration.

pchote commented on mods/ra/rules/misc.yaml:
```diff
@@ -239,6 +245,7 @@ RAILMINE:
 		Dimensions: 2,1
 	EditorTilesetFilter:
 		ExcludeTilesets: INTERIOR
+		Categories: Mine
```
This should be a Civilian building.

pchote commented on mods/ra/rules/defaults.yaml:
```diff
@@ -989,6 +1023,8 @@
 		OccupiesSpace: true
 	HitShape:
+	EditorTilesetFilter:
+		Categories: Mine
```
I don't think it's worth giving these their own category.  How about putting these in System too?

pchote reviewed (commented):
I'm really liking this.

Here's some suggestions to polish the category lists:

rob-v commented:
RA Bridge has empty category, no actors. Remove it from .yaml or should be fixed?
RA has 2 'Resource spawn' actors, TD and TS none. Should we move 'Blossom Tree's to 'Resource spawn' or change all to 'System'?

pchote commented:
RA and TD bridges are currently autogenerated by the map, and shouldn't be placed manually.  I think having a separate Resource Spawn category for all mods is a good idea.  In TS this should include vein holes.

rob-v commented:
Updated.

reaperrr commented on mods/ts/rules/bridges.yaml:
```diff
@@ -41,6 +45,8 @@ CABHUT:
 		Type: Concrete
 	HitShape:
+	EditorTilesetFilter:
+		Categories: Bridge
```
This and all the LOBRDG already inherit it via `^LowBridgeRamp`, so you can remove it here and from the LOBRDGes.

reaperrr commented:
In both RA and TD, in my opinion CAMERA should move to System, and FLARE, mpspawn and waypoint should be added to that, too.

Also, in RA I'd move QUEE and LARVA(E) to Critter and CTFLAG to System.

👍 once that and removing the redundant TS bridge entries is done.

pchote commented:
@reaperrr, @rob-v: would you object to adding this and #13337 to the playtest milestone (and deal with reviews and fixups on the associated deadline)?  It would be nice to be able to add a section on map editor improvements to our playtest annoucement news post.

reaperrr commented:
Fine with me.

rob-v commented:
@reaperrr 

> In both RA and TD, in my opinion CAMERA should move to System, and FLARE, mpspawn and waypoint should be added to that, too.
This is already done as was requested also by @pchote. or not?

reaperrr commented:
> This is already done as was requested also by @pchote. or not?

Sorry, yes, seems I had forgotten to fetch the latest version of the branch. Only the TS bridges need to be updated then.

rob-v commented:
Fixing, please wait with review.

rob-v commented:
> Also, in RA I'd move QUEE and LARVA(E) to Critter and CTFLAG to System.

QUEE Queen Ant has Building trait? I didn't add Category, because it isn't visible in Map editor. @pchote proposed now in IRC to include in Map editor only actors with EditorTilesetFilter (EditorData) trait. Adding Critter could be done, because it probably still won't be displayed for other reasons (as it isn't also now). However this would be inconsistent as I shouldn't add new Category for not visible actors, because such empty category would be in the list.

CTFLAG - is currently Tech building. I can change category, but I think it isn't displayed anyway, so not needed?

rob-v commented:
I didn't see QUEE, because it is visible only in INTERIOR tileset, so I'll add it to Critter.

Commit:
Add RA Actor Categories

Commit:
Add TD Actor Categories

Commit:
Add D2K Actor Categories

Commit:
Add TS Actor Categories

rob-v commented:
Updated Categories as requested, removed No category. To display actor in Map editor, actor must have assigned EditorTilesetFilter with non empty Categories.

reaperrr commented:
@pchote are you fine with the last changes?

pchote commented:
Categories now look much better.  Here are a few final thoughts:
* D2K I'm not sure if it's worth having the frigate, ornithoper accessible in the editor.  They won't work properly when placed on a map.
* D2K having a wall category isn't really worth it (but keep it in the other mods).
* TS hospital, light tower, light posts should move from Building to Civilian Building.
* TS Deployed ICBM probably wants to move somewhere other than Building.
* TS maybe add a dedicated Billboard category? There sure are a lot of those.
* TS maybe merge Gates into Wall?
* TS we seem to be missing all the nod vehicles?
* TD i'm not sure if it's worth having the supply aircraft, a10 accessible in the editor.  They won't work properly when placed on a map.
* TD you have a separate Husk category that contains civilian building husks, but then tree husks are in with the trees.  Please pick a convention (IMO having the civ husks with civs and tree husks with trees would be better, since these are legitimate map props.  "Normal" player-unit husks are temporary actors that probably shouldnt be exposed in the editor in any mod).
* RA i'm not sure if it's worth having the spy plane, or badger variants accessible in the editor.  They won't work properly when placed on a map.
* RA maybe merge Gates into Wall?
* RA explosive barrels and ammo boxes are in with tech buildings?

rob-v commented:
D2K Should I remove frigate, ornithoper?
	D2K Wall moved to Defense.
	TS These actors don't inherit from ^CivBuilding like most others in that .yaml, so added category Civilian building.
	TS Deployed ICBM is in civilian-structures.yaml, so also Civilian building?
	TS Billboards moved to new Billboard category.
	TS Gates moved to Wall.
	TS Nod vehicles are missing also on bleed, though they are added to the panel, any clue?
	TD Should I remove supply aircraft, a10?
	TD I had husks in Husk category, TreeHusk was moved to Tree on your request.
	RA Should I remove spy plane or badger variants?
	RA Gates moved to Wall.
	RA Explosive Barrel and Ammo Box inherit from TechBuilding. To which category they should be moved?
	
Re Aircraft 'They won't work properly when placed on a map.'. If we are not sure, I propose to keep it, it is safer. Next PR can polish categories instead of having to revert it as missing in Map editor due to this PR.
These latest changes are in one commit. After final review I can squash commits if needed.

pchote commented:
For the sake of not dragging out this PR even more, I think it is fine to not remove the aircraft actors.

pchote commented:
TS Deployed ICBM should be in the same category as the normal undeployed ICBM.

pchote commented:
> RA Explosive Barrel and Ammo Box inherit from TechBuilding. To which category they should be moved?

I would put these, the things currently in "Rock" (which includes tank traps on temperate maps), and the boxes / power poles / ice floes that are currently in "Tree" into a "Decoration" category.

cjshmyr commented:
> TS we seem to be missing all the nod vehicles?

Possibly #13492

pchote commented:
> Possibly #13492

This indeed appears to be the case, so we can ignore that here.

Commit:
Polish Actor Categories

rob-v commented:
RA/CNC/TS - Rocks moved to Decoration
RA Explosive Barrel, Ice Floes, Ammo Box moved to Decoration
TS Deployed ICBM moved to Vehicle

reaperrr commented:
My 👍 still stands.

pchote commented:
To save time I have pushed fixes for a couple remaining nits here:
* RA Moved `barl` from Tech building to Decoration.
* RA moved `ctflag` and `flare` from System to Decoration.
* RA moved `boxes01-09` and `utilpol1-2` from Tree to Decoration.

pchote reviewed (approved):
:+1: with my final fixes (pushed here).

@reaperrr if you are happy with those changes then please go ahead and merge this.

reaperrr commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/d8ec9ef2d64bceaaf300a23fcb648f662b541b6b...206180643190b890a66410eeb171946bc3361f6b
